// cspell-trie-lib/dist/lib/TrieBlob/FastTrieBlobInternals.js
import { mergeOptionalWithDefaults } from '../utils/mergeOptionalWithDefaults.js';
export class FastTrieBlobInternals {
    nodes;
    charIndex;
    NodeMaskEOW;
    NodeMaskChildCharIndex;
    NodeChildRefShift;
    isIndexDecoderNeeded;
    info;
    constructor(nodes, charIndex, maskInfo, info) {
        this.nodes = nodes;
        this.charIndex = charIndex;
        const { NodeMaskEOW, NodeMaskChildCharIndex, NodeChildRefShift } = maskInfo;
        this.NodeMaskEOW = NodeMaskEOW;
        this.NodeMaskChildCharIndex = NodeMaskChildCharIndex;
        this.NodeChildRefShift = NodeChildRefShift;
        this.isIndexDecoderNeeded = charIndex.indexContainsMultiByteChars();
        this.info = mergeOptionalWithDefaults(info);
    }
}
export class FastTrieBlobInternalsAndMethods extends FastTrieBlobInternals {
    nodeFindNode;
    nodeFindExact;
    nodeGetChild;
    isForbidden;
    findExact;
    hasForbiddenWords;
    hasCompoundWords;
    hasNonStrictWords;
    constructor(nodes, charIndex, maskInfo, info, trieMethods) {
        super(nodes, charIndex, maskInfo, info);
        this.nodeFindExact = trieMethods.nodeFindExact;
        this.nodeGetChild = trieMethods.nodeGetChild;
        this.isForbidden = trieMethods.isForbidden;
        this.findExact = trieMethods.findExact;
        this.nodeFindNode = trieMethods.nodeFindNode;
        this.hasForbiddenWords = trieMethods.hasForbiddenWords;
        this.hasCompoundWords = trieMethods.hasCompoundWords;
        this.hasNonStrictWords = trieMethods.hasNonStrictWords;
    }
}
/**
 * Sorts the nodes in place if possible.
 * @param nodes
 * @param mask
 * @returns
 */
export function sortNodes(nodes, mask) {
    if (Object.isFrozen(nodes)) {
        assertSorted(nodes, mask);
        return nodes;
    }
    for (let i = 0; i < nodes.length; ++i) {
        let node = nodes[i];
        if (node.length > 2) {
            const isFrozen = Object.isFrozen(node);
            node = isFrozen ? Uint32Array.from(node) : node;
            const nodeInfo = node[0];
            node[0] = 0;
            node.sort((a, b) => (!a ? -1 : !b ? 1 : (a & mask) - (b & mask)));
            // console.log({ i, n: node.map((n, i) => (i ? [n & mask, (n & ~mask) / (mask + 1)] : n)) });
            node[0] = nodeInfo;
            if (isFrozen) {
                nodes[i] = node;
                Object.freeze(node);
            }
        }
    }
    Object.freeze(nodes);
    return nodes;
}
export function assertSorted(nodes, mask) {
    for (let i = 0; i < nodes.length; ++i) {
        const node = nodes[i];
        if (node.length > 2) {
            let last = -1;
            for (let j = 1; j < node.length; ++j) {
                const n = node[j] & mask;
                if (n < last) {
                    throw new Error(`Node ${i} is not sorted. ${last} > ${n}`);
                }
                last = n;
            }
        }
    }
}
