// cspell-trie-lib/dist/lib/TrieBlob/FastTrieBlobBuilder.js
import assert from 'node:assert';
import { mergeOptionalWithDefaults } from '../utils/mergeOptionalWithDefaults.js';
import { assertValidUtf16Character } from '../utils/text.js';
import { CharIndexBuilder } from './CharIndex.js';
import { FastTrieBlob, nodesToJSON } from './FastTrieBlob.js';
import { FastTrieBlobInternals, sortNodes } from './FastTrieBlobInternals.js';
import { resolveMap } from './resolveMap.js';
import { TrieBlob } from './TrieBlob.js';
export class FastTrieBlobBuilder {
    charIndex = new CharIndexBuilder();
    nodes;
    _readonly = false;
    IdxEOW;
    _cursor;
    _options;
    wordToCharacters = (word) => [...word];
    bitMasksInfo;
    constructor(options, bitMasksInfo = FastTrieBlobBuilder.DefaultBitMaskInfo) {
        this._options = mergeOptionalWithDefaults(options);
        this.bitMasksInfo = bitMasksInfo;
        this.nodes = [[0], Object.freeze([FastTrieBlobBuilder.NodeMaskEOW])];
        this.IdxEOW = 1;
    }
    setOptions(options) {
        this._options = mergeOptionalWithDefaults(this.options, options);
        return this.options;
    }
    get options() {
        return this._options;
    }
    wordToUtf8Seq(word) {
        return this.charIndex.wordToUtf8Seq(word);
    }
    letterToUtf8Seq(letter) {
        return this.charIndex.charToUtf8Seq(letter);
    }
    insert(word) {
        this.#assertNotReadonly();
        if (typeof word === 'string') {
            return this._insert(word);
        }
        const words = word;
        for (const w of words) {
            this._insert(w);
        }
        return this;
    }
    getCursor() {
        this.#assertNotReadonly();
        this._cursor ??= this.createCursor();
        return this._cursor;
    }
    createCursor() {
        const NodeChildRefShift = this.bitMasksInfo.NodeChildRefShift;
        const NodeMaskEOW = this.bitMasksInfo.NodeMaskEOW;
        const LetterMask = this.bitMasksInfo.NodeMaskChildCharIndex;
        const refNodes = [0, 1];
        function childPos(node, letterIdx) {
            for (let i = 1; i < node.length; ++i) {
                if ((node[i] & LetterMask) === letterIdx) {
                    return i;
                }
            }
            return 0;
        }
        assert(this.nodes.length === 2);
        const eow = 1;
        const eowShifted = eow << NodeChildRefShift;
        const nodes = this.nodes;
        const stack = [{ nodeIdx: 0, pos: 0, pDepth: -1 }];
        let nodeIdx = 0;
        let depth = 0;
        const insertChar = (char) => {
            if (!nodes[nodeIdx]) {
                refNodes.push(nodeIdx);
            }
            // console.warn('insertChar %o', { nodeIdx, depth, char });
            const pDepth = depth;
            const utf8Seq = this.letterToUtf8Seq(char);
            for (let i = 0; i < utf8Seq.length; ++i) {
                insertCharIndexes(utf8Seq[i], pDepth);
            }
        };
        /**
         * A single character can result in multiple nodes being created
         * because it takes multiple bytes to represent a character.
         * @param seq - partial character index.
         */
        const insertCharIndexes = (seq, pDepth) => {
            // console.warn('i %o at %o', char, nodeIdx);
            if (nodes[nodeIdx] && Object.isFrozen(nodes[nodeIdx])) {
                nodeIdx = nodes.push([...nodes[nodeIdx]]) - 1;
                // fix parent
                const { pos, nodeIdx: pNodeIdx } = stack[depth];
                const pNode = nodes[pNodeIdx];
                // console.warn('fix parent %o', { pNode, pos, pNodeIdx });
                pNode[pos] = (pNode[pos] & LetterMask) | (nodeIdx << NodeChildRefShift);
            }
            const node = nodes[nodeIdx] || [0];
            nodes[nodeIdx] = node;
            const hasIdx = childPos(node, seq);
            const childIdx = hasIdx ? node[hasIdx] >>> NodeChildRefShift : nodes.length;
            const pos = hasIdx || node.push((childIdx << NodeChildRefShift) | seq) - 1;
            ++depth;
            const s = stack[depth];
            if (s) {
                s.nodeIdx = nodeIdx;
                s.pos = pos;
                s.pDepth = pDepth;
            }
            else {
                stack[depth] = { nodeIdx, pos, pDepth };
            }
            nodeIdx = childIdx;
        };
        const markEOW = () => {
            // console.warn('$');
            if (nodeIdx === eow)
                return;
            const node = nodes[nodeIdx];
            if (!node) {
                // no children, set the parent to point to the common EOW.
                const { pos, nodeIdx: pNodeIdx } = stack[depth];
                const pNode = nodes[pNodeIdx];
                pNode[pos] = (pNode[pos] & LetterMask) | eowShifted;
            }
            else {
                nodes[nodeIdx] = node;
                node[0] |= NodeMaskEOW;
            }
            nodeIdx = eow;
        };
        const reference = (refId) => {
            const refNodeIdx = refNodes[refId];
            assert(refNodeIdx !== undefined);
            // console.warn('r %o', { refId, nodeIdx, refNodeIdx, depth });
            assert(nodes[nodeIdx] === undefined);
            assert(nodes[refNodeIdx]);
            Object.freeze(nodes[refNodeIdx]);
            const s = stack[depth];
            nodeIdx = s.nodeIdx;
            const pos = s.pos;
            const node = nodes[nodeIdx];
            node[pos] = (refNodeIdx << NodeChildRefShift) | (node[pos] & LetterMask);
        };
        const backStep = (num) => {
            if (!num)
                return;
            // console.warn('<< %o', num);
            assert(num <= depth && num > 0);
            for (let n = num; n > 0; --n) {
                depth = stack[depth].pDepth;
            }
            nodeIdx = stack[depth + 1].nodeIdx;
        };
        const c = {
            insertChar,
            markEOW,
            reference,
            backStep,
        };
        return c;
    }
    _insert(word) {
        word = word.trim();
        if (!word)
            return this;
        const NodeMaskChildCharIndex = this.bitMasksInfo.NodeMaskChildCharIndex;
        const NodeChildRefShift = this.bitMasksInfo.NodeChildRefShift;
        const NodeMaskEOW = this.bitMasksInfo.NodeMaskEOW;
        const IdxEOW = this.IdxEOW;
        const nodes = this.nodes;
        const utf8Seq = this.wordToUtf8Seq(word);
        const len = utf8Seq.length;
        let nodeIdx = 0;
        for (let p = 0; p < len; ++p) {
            const seq = utf8Seq[p];
            const node = nodes[nodeIdx];
            const count = node.length;
            let i = count - 1;
            for (; i > 0; --i) {
                if ((node[i] & NodeMaskChildCharIndex) === seq) {
                    break;
                }
            }
            if (i > 0) {
                nodeIdx = node[i] >>> NodeChildRefShift;
                if (nodeIdx === 1 && p < len - 1) {
                    nodeIdx = this.nodes.push([NodeMaskEOW]) - 1;
                    node[i] = (nodeIdx << NodeChildRefShift) | seq;
                }
                continue;
            }
            // Not found, add a new node if it isn't the end of the word.
            nodeIdx = p < len - 1 ? this.nodes.push([0]) - 1 : IdxEOW;
            node.push((nodeIdx << NodeChildRefShift) | seq);
        }
        if (nodeIdx > 1) {
            // Make sure the EOW is set
            const node = nodes[nodeIdx];
            node[0] |= NodeMaskEOW;
        }
        return this;
    }
    has(word) {
        const NodeMaskChildCharIndex = this.bitMasksInfo.NodeMaskChildCharIndex;
        const NodeChildRefShift = this.bitMasksInfo.NodeChildRefShift;
        const NodeMaskEOW = this.bitMasksInfo.NodeMaskEOW;
        const nodes = this.nodes;
        const charIndexes = this.wordToUtf8Seq(word);
        const len = charIndexes.length;
        let nodeIdx = 0;
        let node = nodes[nodeIdx];
        for (let p = 0; p < len; ++p, node = nodes[nodeIdx]) {
            const letterIdx = charIndexes[p];
            const count = node.length;
            let i = count - 1;
            for (; i > 0; --i) {
                if ((node[i] & NodeMaskChildCharIndex) === letterIdx) {
                    break;
                }
            }
            if (i < 1)
                return false;
            nodeIdx = node[i] >>> NodeChildRefShift;
        }
        return !!(node[0] & NodeMaskEOW);
    }
    isReadonly() {
        return this._readonly;
    }
    freeze() {
        this._readonly = true;
        return this;
    }
    build() {
        this._cursor = undefined;
        this._readonly = true;
        this.freeze();
        return FastTrieBlob.create(new FastTrieBlobInternals(sortNodes(this.nodes.map((n) => Uint32Array.from(n)), this.bitMasksInfo.NodeMaskChildCharIndex), this.charIndex.build(), this.bitMasksInfo, this.options));
    }
    toJSON() {
        return {
            options: this.options,
            nodes: nodesToJSON(this.nodes.map((n) => Uint32Array.from(n))),
        };
    }
    #assertNotReadonly() {
        assert(!this.isReadonly(), 'FastTrieBlobBuilder is readonly');
    }
    static fromWordList(words, options) {
        const ft = new FastTrieBlobBuilder(options);
        return ft.insert(words).build();
    }
    static fromTrieRoot(root) {
        const bitMasksInfo = FastTrieBlobBuilder.DefaultBitMaskInfo;
        const NodeChildRefShift = bitMasksInfo.NodeChildRefShift;
        const NodeCharIndexMask = bitMasksInfo.NodeMaskChildCharIndex;
        const NodeMaskEOW = bitMasksInfo.NodeMaskEOW;
        const tf = new FastTrieBlobBuilder(undefined, bitMasksInfo);
        const IdxEOW = tf.IdxEOW;
        const known = new Map([[root, 0]]);
        function resolveNode(n) {
            if (n.f && !n.c)
                return IdxEOW;
            const node = [n.f ? NodeMaskEOW : 0];
            return tf.nodes.push(node) - 1;
        }
        function walk(n) {
            const found = known.get(n);
            if (found)
                return found;
            const nodeIdx = resolveMap(known, n, resolveNode);
            const node = tf.nodes[nodeIdx];
            if (!n.c)
                return nodeIdx;
            const children = Object.entries(n.c);
            for (let p = 0; p < children.length; ++p) {
                const [char, childNode] = children[p];
                addCharToNode(node, char, childNode);
            }
            return nodeIdx;
        }
        function resolveChild(node, charIndex) {
            let i = 1;
            for (i = 1; i < node.length && (node[i] & NodeCharIndexMask) !== charIndex; ++i) {
                // empty
            }
            return i;
        }
        function addCharToNode(node, char, n) {
            const indexSeq = tf.letterToUtf8Seq(char);
            assertValidUtf16Character(char);
            // console.error('addCharToNode %o', { char, indexSeq });
            for (const idx of indexSeq.slice(0, -1)) {
                const pos = resolveChild(node, idx);
                if (pos < node.length) {
                    node = tf.nodes[node[pos] >>> NodeChildRefShift];
                }
                else {
                    const next = [0];
                    const nodeIdx = tf.nodes.push(next) - 1;
                    node[pos] = (nodeIdx << NodeChildRefShift) | idx;
                    node = next;
                }
            }
            const letterIdx = indexSeq[indexSeq.length - 1];
            const i = node.push(letterIdx) - 1;
            node[i] = (walk(n) << NodeChildRefShift) | letterIdx;
        }
        walk(root);
        return tf.build();
    }
    static NodeMaskEOW = TrieBlob.NodeMaskEOW;
    static NodeChildRefShift = TrieBlob.NodeChildRefShift;
    static NodeMaskChildCharIndex = TrieBlob.NodeMaskChildCharIndex;
    static DefaultBitMaskInfo = {
        NodeMaskEOW: FastTrieBlobBuilder.NodeMaskEOW,
        NodeMaskChildCharIndex: FastTrieBlobBuilder.NodeMaskChildCharIndex,
        NodeChildRefShift: FastTrieBlobBuilder.NodeChildRefShift,
    };
}
